### ISSUE TITLE
incorrectly branching into cheatcode call during cse summarization

### ISSUE THREAD

**Issue by ovatman** [runtimeverification/kontrol#822]
When applying cse kontrol produces a `#cheatcode_call` branch due to a possible misinterpretation of a symoblic variable when applying `#checkCall`.
To reproduce, perform the following, using kontrol 1.0.25
1. Use the UniswapV2 variant from [here](https://github.com/Pi-Squared-Inc/solidity-demo-semantics/blob/main/test/examples/swaps/UniswapV2SwapRenamed.sol); a part of the Pi^2 solidity semantic development project. This file contains multiple contracts.
2. Initialize a foundry project and place the test contract in a separate test directory. In my case all the other contracts reside in a single file under `src` and the `uniswapV2SwapTest` was under `test`
3. `forge test` and `kontrol build`
4. We will try to get a summary of the `uniswapV2Pair.sync()` function so run `kontrol prove --match-test 'uniswapV2Pair.sync()' --break-on-calls --cse` 

Should fail as follows:
```
  Failure reason:
    Matching failed.
    The following cells failed matching individually (antecedent #Implies consequent):
    K_CELL: #cheatcode_error selector ( "balanceOf(address)" ) #buf ( 32 , C_UNISWAPV2PAIR_ID:Int )
    ~> #cheatcode_return 128 32
    ~> #pc [ CALL ]
    ~> #execute #Implies #halt ~> .K
    ~> CONTINUATION:K
```

6. After failure, issue `kontrol show 'uniswapV2Pair.sync()'`

Following is the related part from kcfg

```
├─ 5
│   k: #checkCall C_UNISWAPV2PAIR_ID:Int 0 ~> #call C_UNISWAPV2PAIR_ID:Int C_UNISWAPV2P ...
│   pc: 2313
│   callDepth: CALLDEPTH_CELL:Int
│   statusCode: STATUSCODE:StatusCode
│   src: src/UniswapV2Swap.sol:290:290
│   method: src%uniswapV2Pair.sync()
┃
┃ (1 step)
┣━━┓
┃  │
┃  ├─ 6
┃  │   k: #cheatcode_call selector ( "balanceOf(address)" ) #buf ( 32 , C_UNISWAPV2PAIR_ID ...
┃  │   pc: 2313
┃  │   callDepth: CALLDEPTH_CELL:Int
┃  │   statusCode: STATUSCODE:StatusCode
┃  │   src: src/UniswapV2Swap.sol:290:290
┃  │   method: src%uniswapV2Pair.sync()
┃  │
┃  │  (1 step)
┃  └─ 8 (stuck, leaf)
┃      k: #cheatcode_error selector ( "balanceOf(address)" ) #buf ( 32 , C_UNISWAPV2PAIR_I ...
┃      pc: 2313
┃      callDepth: CALLDEPTH_CELL:Int
┃      statusCode: CHEATCODE_UNIMPLEMENTED
┃      src: src/UniswapV2Swap.sol:290:290
┃      method: src%uniswapV2Pair.sync()
```

**Comment by PetarMax** [comment]
This is strange, because (I think) this means that `C_UNISWAPV2PAIR_ID` does not have the constraint that it's different from the cheatcode account address. We normally put this constraint in. @ovatman, what is the full configuration you see in node 5?

**Comment by PetarMax** [comment]
Ok, thank you, could you please give me the configuration of node 6 as well?

**Comment by PetarMax** [comment]
```
#And ( { true #Equals ( notBool C_UNISWAPV2PAIR_ID:Int ==Int 645326474426547203313410069153905908525362434349 ) }
```
is there, which should prevent the branching...

**Comment by PetarMax** [comment]
@palinatolmach When summarising a function, if one of the parameters is an address, I think that we need to assume that it does not equal the cheatcode account address.

**Comment by palinatolmach** [comment]
Agreed @PetarMax, I'll open an issue and a PR for it today. Should we do it for an `address`/`contract` function parameter, or an `address`/`contract` field, or both? Probably both?

**Comment by PetarMax** [comment]
Both, I think.

**Comment by palinatolmach** [comment]
Thanks! Opened and assigned myself: https://github.com/runtimeverification/kontrol/issues/827.

**Comment by palinatolmach** [comment]
Closed in favor of https://github.com/runtimeverification/kontrol/issues/827.
